File: baboon-compiler/src/main/resources/baboon-runtime/cs/BaboonServiceWiring.cs
#nullable enable

using System;
using System.Collections.Generic;

// ReSharper disable UnusedTypeParameter
// ReSharper disable CheckNamespace
// ReSharper disable UnusedAutoPropertyAccessor.Global
// ReSharper disable MemberCanBeProtected.Global
// ReSharper disable MemberCanBePrivate.Global
// ReSharper disable ConvertToPrimaryConstructor
// ReSharper disable UnusedMember.Global
// ReSharper disable UnusedMemberInSuper.Global
// ReSharper disable UseCollectionExpression
// ReSharper disable ReplaceAutoPropertyWithComputedProperty
// ReSharper disable ArrangeNamespaceBody
// ReSharper disable UnusedType.Global
// ReSharper disable InconsistentNaming
// ReSharper disable ClassCanBeSealed.Global

namespace Baboon.Runtime.Shared
{
    public sealed record BaboonMethodId(string ServiceName, string MethodName);

    public abstract record BaboonWiringError
    {
        private BaboonWiringError()
        {
        }

        public sealed record NoMatchingMethod(BaboonMethodId Method) : BaboonWiringError;

        public sealed record NoMatchingService(BaboonMethodId Method) : BaboonWiringError;

        public sealed record DuplicateService(string ServiceName) : BaboonWiringError;

        public sealed record DecoderFailed(BaboonMethodId Method, Exception Exception) : BaboonWiringError;

        public sealed record EncoderFailed(BaboonMethodId Method, Exception Exception) : BaboonWiringError;

        public sealed record CallFailed(BaboonMethodId Method, object? DomainError) : BaboonWiringError;
    }

    public class BaboonWiringException : Exception
    {
        public BaboonWiringError Error { get; }

        public BaboonWiringException(BaboonWiringError error) : base(error.ToString())
        {
            Error = error;
        }
    }

    // --- Service muxers ---
    //
    // Cross-domain composable dispatch. A muxer holds a set of services from
    // any model(s) and routes an `(method, data, ctx)` call to the right one
    // by `method.ServiceName`. The R type parameter encodes the return shape
    // so the same class supports both sync (R = string / byte[]) and
    // hypothetical async (R = Task<string> / Task<byte[]>) generated
    // services. The C# wiring generator currently emits sync code only, so
    // the per-service wrapper classes pick R = string / byte[] (or the
    // service-result-container-wrapped variant in errors mode), but the
    // muxer itself stays parametric for forward-compat with potential
    // future Task-based emission.

    public interface IBaboonJsonService<out R>
    {
        string ServiceName { get; }
        R Invoke(BaboonMethodId method, string data, BaboonCodecContext ctx);
    }

    public interface IBaboonUebaService<out R>
    {
        string ServiceName { get; }
        R Invoke(BaboonMethodId method, byte[] data, BaboonCodecContext ctx);
    }

    public sealed class JsonMuxer<R>
    {
        private readonly Dictionary<string, IBaboonJsonService<R>> _table = new Dictionary<string, IBaboonJsonService<R>>();

        public JsonMuxer(params IBaboonJsonService<R>[] services)
        {
            foreach (var s in services)
            {
                Register(s);
            }
        }

        public void Register(IBaboonJsonService<R> service)
        {
            if (_table.ContainsKey(service.ServiceName))
            {
                throw new BaboonWiringException(new BaboonWiringError.DuplicateService(service.ServiceName));
            }
            _table[service.ServiceName] = service;
        }

        public R Invoke(BaboonMethodId method, string data, BaboonCodecContext ctx)
        {
            if (!_table.TryGetValue(method.ServiceName, out var service))
            {
                throw new BaboonWiringException(new BaboonWiringError.NoMatchingService(method));
            }
            return service.Invoke(method, data, ctx);
        }

        public IReadOnlyCollection<string> ServiceNames()
        {
            return _table.Keys;
        }
    }

    public sealed class UebaMuxer<R>
    {
        private readonly Dictionary<string, IBaboonUebaService<R>> _table = new Dictionary<string, IBaboonUebaService<R>>();

        public UebaMuxer(params IBaboonUebaService<R>[] services)
        {
            foreach (var s in services)
            {
                Register(s);
            }
        }

        public void Register(IBaboonUebaService<R> service)
        {
            if (_table.ContainsKey(service.ServiceName))
            {
                throw new BaboonWiringException(new BaboonWiringError.DuplicateService(service.ServiceName));
            }
            _table[service.ServiceName] = service;
        }

        public R Invoke(BaboonMethodId method, byte[] data, BaboonCodecContext ctx)
        {
            if (!_table.TryGetValue(method.ServiceName, out var service))
            {
                throw new BaboonWiringException(new BaboonWiringError.NoMatchingService(method));
            }
            return service.Invoke(method, data, ctx);
        }

        public IReadOnlyCollection<string> ServiceNames()
        {
            return _table.Keys;
        }
    }

    // Context-carrying variants, emitted when a service.context mode
    // (`abstract` or `type`) is active. The service context `Ctx` is supplied
    // per `Invoke` (alongside the codec context) instead of being baked into
    // the wrapper at construction time. The context-free interfaces above are
    // left untouched so `--service-context-mode none` output (and the
    // service-acceptance matrix) is byte-identical.
    public interface IBaboonJsonServiceCtx<Ctx, out R>
    {
        string ServiceName { get; }
        R Invoke(BaboonMethodId method, string data, Ctx ctx, BaboonCodecContext codecCtx);
    }

    public interface IBaboonUebaServiceCtx<Ctx, out R>
    {
        string ServiceName { get; }
        R Invoke(BaboonMethodId method, byte[] data, Ctx ctx, BaboonCodecContext codecCtx);
    }

    public sealed class JsonMuxerCtx<Ctx, R>
    {
        private readonly Dictionary<string, IBaboonJsonServiceCtx<Ctx, R>> _table = new Dictionary<string, IBaboonJsonServiceCtx<Ctx, R>>();

        public JsonMuxerCtx(params IBaboonJsonServiceCtx<Ctx, R>[] services)
        {
            foreach (var s in services)
            {
                Register(s);
            }
        }

        public void Register(IBaboonJsonServiceCtx<Ctx, R> service)
        {
            if (_table.ContainsKey(service.ServiceName))
            {
                throw new BaboonWiringException(new BaboonWiringError.DuplicateService(service.ServiceName));
            }
            _table[service.ServiceName] = service;
        }

        public R Invoke(BaboonMethodId method, string data, Ctx ctx, BaboonCodecContext codecCtx)
        {
            if (!_table.TryGetValue(method.ServiceName, out var service))
            {
                throw new BaboonWiringException(new BaboonWiringError.NoMatchingService(method));
            }
            return service.Invoke(method, data, ctx, codecCtx);
        }

        public IReadOnlyCollection<string> ServiceNames()
        {
            return _table.Keys;
        }
    }

    public sealed class UebaMuxerCtx<Ctx, R>
    {
        private readonly Dictionary<string, IBaboonUebaServiceCtx<Ctx, R>> _table = new Dictionary<string, IBaboonUebaServiceCtx<Ctx, R>>();

        public UebaMuxerCtx(params IBaboonUebaServiceCtx<Ctx, R>[] services)
        {
            foreach (var s in services)
            {
                Register(s);
            }
        }

        public void Register(IBaboonUebaServiceCtx<Ctx, R> service)
        {
            if (_table.ContainsKey(service.ServiceName))
            {
                throw new BaboonWiringException(new BaboonWiringError.DuplicateService(service.ServiceName));
            }
            _table[service.ServiceName] = service;
        }

        public R Invoke(BaboonMethodId method, byte[] data, Ctx ctx, BaboonCodecContext codecCtx)
        {
            if (!_table.TryGetValue(method.ServiceName, out var service))
            {
                throw new BaboonWiringException(new BaboonWiringError.NoMatchingService(method));
            }
            return service.Invoke(method, data, ctx, codecCtx);
        }

        public IReadOnlyCollection<string> ServiceNames()
        {
            return _table.Keys;
        }
    }
}

File: baboon-compiler/src/main/resources/baboon-runtime/python/baboon_service_wiring.py
from dataclasses import dataclass
from typing import Any, Awaitable, Dict, Generic, Protocol, TypeVar, Union, runtime_checkable


@dataclass(frozen=True)
class BaboonMethodId:
    service_name: str
    method_name: str


class BaboonWiringError:
    pass


class NoMatchingMethod(BaboonWiringError):
    def __init__(self, method: BaboonMethodId):
        self.method = method

    def __repr__(self):
        return f"NoMatchingMethod({self.method})"


class NoMatchingService(BaboonWiringError):
    def __init__(self, method: BaboonMethodId):
        self.method = method

    def __repr__(self):
        return f"NoMatchingService({self.method})"


class DuplicateService(BaboonWiringError):
    def __init__(self, service_name: str):
        self.service_name = service_name

    def __repr__(self):
        return f"DuplicateService({self.service_name!r})"


class DecoderFailed(BaboonWiringError):
    def __init__(self, method: BaboonMethodId, exception: Exception):
        self.method = method
        self.exception = exception

    def __repr__(self):
        return f"DecoderFailed({self.method}, {self.exception})"


class EncoderFailed(BaboonWiringError):
    def __init__(self, method: BaboonMethodId, exception: Exception):
        self.method = method
        self.exception = exception

    def __repr__(self):
        return f"EncoderFailed({self.method}, {self.exception})"


class CallFailed(BaboonWiringError):
    def __init__(self, method: BaboonMethodId, domain_error: Any):
        self.method = method
        self.domain_error = domain_error

    def __repr__(self):
        return f"CallFailed({self.method}, {self.domain_error})"


class BaboonWiringException(Exception):
    def __init__(self, error: BaboonWiringError):
        super().__init__(str(error))
        self.error = error


class BaboonEither:
    pass


class BaboonLeft(BaboonEither):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"BaboonLeft({self.value!r})"


class BaboonRight(BaboonEither):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"BaboonRight({self.value!r})"


# --- Service muxers ---
#
# Cross-domain composable dispatch. A muxer holds a set of services from any
# model(s) and routes an `(method, data, ctx)` call to the right one by
# `method.service_name`. The R type parameter encodes the return shape so the
# same class supports both sync and async generated services — for sync code
# use `JsonMuxer[str]` / `UebaMuxer[bytes]`; for async code use
# `JsonMuxer[Awaitable[str]]` / `UebaMuxer[Awaitable[bytes]]`. The per-service
# wrapper classes emitted alongside `invoke_json_X` / `invoke_ueba_X` carry
# the matching parameterisation. The muxer is intentionally a single class
# rather than two (sync / async) — generic in R via PEP 484 type-vars; the
# wrapper class's `invoke` annotation picks the concrete return type.
#
# Python `Protocol` (PEP 544) gives us structural typing so the generated
# wrapper classes can satisfy `IBaboonJsonService` / `IBaboonUebaService`
# without an explicit `class X(IBaboonJsonService)` declaration — purely by
# having the right attributes.

R = TypeVar("R")


@runtime_checkable
class IBaboonJsonService(Protocol[R]):
    service_name: str

    def invoke(self, method: BaboonMethodId, data: str, ctx) -> R: ...


@runtime_checkable
class IBaboonUebaService(Protocol[R]):
    service_name: str

    def invoke(self, method: BaboonMethodId, data: bytes, ctx) -> R: ...


class JsonMuxer(Generic[R]):
    """Cross-domain JSON dispatcher. Routes `(method, data, ctx)` to a registered
    service whose `service_name` matches `method.service_name`. Strict fail-fast
    on duplicate registration or unknown service.
    """

    def __init__(self, *services: IBaboonJsonService[R]) -> None:
        self._table: Dict[str, IBaboonJsonService[R]] = {}
        for s in services:
            self.register(s)

    def register(self, service: IBaboonJsonService[R]) -> None:
        if service.service_name in self._table:
            raise BaboonWiringException(DuplicateService(service.service_name))
        self._table[service.service_name] = service

    def invoke(self, method: BaboonMethodId, data: str, ctx) -> R:
        service = self._table.get(method.service_name)
        if service is None:
            raise BaboonWiringException(NoMatchingService(method))
        return service.invoke(method, data, ctx)

    def service_names(self) -> list:
        return list(self._table.keys())


class UebaMuxer(Generic[R]):
    """Cross-domain UEBA dispatcher. Routes `(method, data, ctx)` to a registered
    service whose `service_name` matches `method.service_name`. Strict fail-fast
    on duplicate registration or unknown service.
    """

    def __init__(self, *services: IBaboonUebaService[R]) -> None:
        self._table: Dict[str, IBaboonUebaService[R]] = {}
        for s in services:
            self.register(s)

    def register(self, service: IBaboonUebaService[R]) -> None:
        if service.service_name in self._table:
            raise BaboonWiringException(DuplicateService(service.service_name))
        self._table[service.service_name] = service

    def invoke(self, method: BaboonMethodId, data: bytes, ctx) -> R:
        service = self._table.get(method.service_name)
        if service is None:
            raise BaboonWiringException(NoMatchingService(method))
        return service.invoke(method, data, ctx)

    def service_names(self) -> list:
        return list(self._table.keys())
